core/server/worker/src/main/java/alluxio/worker/DefaultAlluxioWorker.java

```diff
@@ -95,35 +88,33 @@
   private long mStartTimeMs;
 
   /**
-   * The worker ID for this worker. This is set when the block worker is initialized and may be
-   * updated by the block sync thread if the master requests re-registration.
-   */
-  private AtomicReference<Long> mWorkerId;
-
-  /**
    * Creates a new instance of {@link DefaultAlluxioWorker}.
    */
   public DefaultAlluxioWorker() {
     try {
-      mWorkerId = new AtomicReference<>();
       mStartTimeMs = System.currentTimeMillis();
-      mBlockWorker = new DefaultBlockWorker(mWorkerId);
-      mFileSystemWorker = new DefaultFileSystemWorker(mBlockWorker, mWorkerId);
-
-      mAdditionalWorkers = new ArrayList<>();
-      List<? extends Worker> workers = Lists.newArrayList(mBlockWorker, mFileSystemWorker);
-      for (WorkerFactory factory : ServerUtils.getWorkerServiceLoader()) {
-        Worker worker = factory.create(workers);
-        if (worker != null) {
-          mAdditionalWorkers.add(worker);
-        }
+      mRegistry = new Registry<>();
+      List<Callable<Void>> callables = new ArrayList<>();
+      for (final WorkerFactory factory : ServiceUtils.getWorkerServiceLoader()) {
+        callables.add(new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            factory.create(mRegistry);
```

Comment:
the registry is passed to the constructor so that the instance can add itself to the registry ... passing additional arguments that can be derived from the registry seems redundant

as an alternative, we can add the instance to the registry in the factory, which is okay as long as no one except for the factory creates instances of the services directly (which is currently true with the exception of a couple of tests) ... what are your thoughts?